### PR TITLE
dev event_group_category

### PR DESCRIPTION
**Pull request by Raulnet**

### CODE REVIEW

File: src/DataFixtures/ORM/LoadEventCategoryData.php [hunk 1]
```diff
@@ -21,16 +21,37 @@ class LoadEventCategoryData implements FixtureInterface
         'CE010' => 'Marche',
         'CE011' => 'Support party',
     ];
-    const HIDDEN_CATEGORY_NAME = 'Catégorie masquée';
+
+    public const EVENT_CATEGORIES_GROUPED = [
```
Comment (ottaviano) [inline src/DataFixtures/ORM/LoadEventCategoryData.php]: maybe `EVENT_CATEGORY_GROUPE_MAP`?

File: src/DataFixtures/ORM/LoadEventCategoryData.php [hunk 2]
```diff
@@ -21,16 +21,37 @@ class LoadEventCategoryData implements FixtureInterface
...
+        'ancrage local' => 'event-group-category-1',
+        'projets citoyens' => 'event-group-category-1',
+        'Un An' => 'event-group-category-1',
+        'Débat' => 'event-group-category-2',
+    ];
+
+    public const HIDDEN_CATEGORY_NAME = 'Catégorie masquée';
```
Comment (ottaviano) [inline src/DataFixtures/ORM/LoadEventCategoryData.php]: why do not include this category into the previous const array?

File: src/Entity/EventCategory.php [hunk 3]
```diff
@@ -22,4 +22,19 @@
  */
 class EventCategory extends BaseEventCategory
 {
+    /**
+     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\EventGroupCategory", inversedBy="eventCategories")
+     * @ORM\JoinColumn(name="event_group_category_id", referencedColumnName="id", nullable=true)
```
Comment (ottaviano) [inline src/Entity/EventCategory.php]: I think you can avoid this annotation, by default the column will be nullable

File: src/Entity/EventGroupCategory.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+use Doctrine\Common\Collections\ArrayCollection;
+use Algolia\AlgoliaSearchBundle\Mapping\Annotation as Algolia;
+use Doctrine\Common\Collections\Collection;
+use Doctrine\ORM\Mapping as ORM;
+use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;
+
+/**
+ * @ORM\Entity(repositoryClass="AppBundle\Repository\EventGroupCategoryRepository")
+ * @ORM\Table(
+ *     name="events_groups_categories",
```
Comment (ottaviano) [inline src/Entity/EventGroupCategory.php]: you can avoid the name attribute and late doctrine manage it

File: src/Entity/EventGroupCategory.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            $this->eventCategories[] = $eventCategory;
+            $eventCategory->setEventGroupCategory($this);
+        }
+
+        return $this;
+    }
+
+    public function removeEventCategory(EventCategory $eventCategory): self
+    {
+        if ($this->eventCategories->contains($eventCategory)) {
```
Comment (ottaviano) [inline src/Entity/EventGroupCategory.php]: `removeElement` method does this check for you

File: src/Admin/EventGroupCategoryAdmin.php [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+class EventGroupCategoryAdmin extends EventCategoryAdmin
+{
+    protected function configureFormFields(FormMapper $formMapper)
+    {
+        $formMapper
+            ->add('name', TextType::class, [
+                'label' => 'Nom',
+            ])
+            ->add('slug', null, [
```
Comment (ottaviano) [inline src/Admin/EventGroupCategoryAdmin.php]: why do we need to have slug & status fields for this Entity? maybe we can keep it simpler with only name property?

File: src/Controller/EnMarche/SearchController.php [hunk 7]
```diff
@@ -51,8 +51,19 @@ public function searchEventsAction(Request $request, string $slug = null)
             $errors[] = $this->get('translator')->trans('search.geocoding.exception');
         }
 
+        $eventCategories = [];
+        /** @var EventCategory[] $categories */
+        $categories = $this->getDoctrine()->getRepository(EventCategory::class)->findAllEnabledOrderedByName();
+        foreach ($categories as $category) {
```
Comment (ottaviano) [inline src/Controller/EnMarche/SearchController.php]: you should move this in the repository

File: src/Controller/EnMarche/SearchController.php [hunk 8]
```diff
@@ -51,8 +51,19 @@ public function searchEventsAction(Request $request, string $slug = null)
...
+            if (null !== $category->getEventGroupCategory()) {
+                $eventCategories[$category->getEventGroupCategory()->getSlug()] = $category->getEventGroupCategory();
+            } else {
+                $eventCategories[$category->getSlug()] = $category;
+            }
+        }
+
         return $this->render('search/search_events.html.twig', [
-            'event_categories' => $this->getDoctrine()->getRepository(EventCategory::class)->findAllEnabledOrderedByName(),
+            'event_categories' => $eventCategories,
```
Comment (ottaviano) [inline src/Controller/EnMarche/SearchController.php]: I think you should order by Group name the list

File: templates/form_theme_admin.html.twig [hunk 9]
```diff
@@ -1,4 +1,4 @@
-{% extends '@SonataAdmin/Form/form_admin_fields.html.twig' %}
+{% extends '@SonataDoctrineORMAdmin/Form/form_admin_fields.html.twig' %}
```
Comment (ottaviano) [inline templates/form_theme_admin.html.twig]: why this change?

Comment (Raulnet) [reply templates/form_theme_admin.html.twig]: Everything. SonataDoctrineAdmin Overide SonataAdmin and give  more option for oneToMAny etc...

File: src/DataFixtures/ORM/LoadEventCategoryData.php (hunk 2 above)
Comment (Raulnet) [reply src/DataFixtures/ORM/LoadEventCategoryData.php]: add after refacto

File: src/DataFixtures/ORM/LoadEventCategoryData.php (hunk 1 above)
Comment (Raulnet) [reply src/DataFixtures/ORM/LoadEventCategoryData.php]: Loader data was edited

File: src/Entity/EventCategory.php (hunk 3 above)
Comment (Raulnet) [reply src/Entity/EventCategory.php]: removed

File: src/Entity/EventGroupCategory.php (hunk 4 above)
Comment (Raulnet) [reply src/Entity/EventGroupCategory.php]: removed

File: templates/search/_filters.html.twig [hunk 10]
```diff
@@ -4,11 +4,20 @@
...
+                        class="form">
+                    <option value=""{% if not search.eventCategory %} selected="selected"{% endif %}>Tous les
+                        événements
+                    </option>
                     <option value="{{ constant('TYPE_CITIZEN_ACTIONS', search) }}">Actions citoyennes</option>
                     {% for category in event_categories %}
-                        <option value="{{ category.id }}"{% if search.eventCategory == category.id %} selected="selected"{% endif %}>{{ category.name }}</option>
+                        <optgroup label="{{ category.name }}">
+                            {% for category in category.eventCategories %}
+                                {% if category.status == 'ENABLED' %}
```
Comment (ottaviano) [inline templates/search/_filters.html.twig]: IMO it's not a good place for this check

File: src/Form/EventGroupCategoryType.php [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+class EventGroupCategoryType extends AbstractType
+{
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'class' => EventCategory::class,
+            'choice_label' => 'name',
+            'query_builder' => function (EntityRepository $er) {
+                return $er->createQueryBuilder('e')
```
Comment (ottaviano) [inline src/Form/EventGroupCategoryType.php]: `e` for `category`? 🤔

File: src/Form/EventGroupCategoryType.php [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+                    ->leftJoin('e.eventGroupCategory', 'egc')
+                    ->where('e.status = :status')
+                    ->andWhere('egc.status = :egc_status')
+                    ->orWhere('e.eventGroupCategory IS NULL')
```
Comment (ottaviano) [inline src/Form/EventGroupCategoryType.php]: can you review this query please?

File: src/Form/EventGroupCategoryType.php (hunk 11 above)
Comment (Raulnet) [reply src/Form/EventGroupCategoryType.php]: e for event .. I change for ec

File: src/Entity/EventCategory.php [hunk 13]
```diff
@@ -22,4 +22,19 @@
  */
 class EventCategory extends BaseEventCategory
 {
+    /**
+     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\EventGroupCategory", inversedBy="eventCategories")
+     * @ORM\JoinColumn(name="event_group_category_id", referencedColumnName="id")
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply src/Entity/EventCategory.php]: set at nullable=false and migration edited

File: src/Form/EventGroupCategoryType.php (hunk 12 above)
Comment (Raulnet) [reply src/Form/EventGroupCategoryType.php]: reviewed && fixed

File: templates/search/_filters.html.twig [hunk 14]
```diff
@@ -4,11 +4,20 @@
...
             <div class="search__bar__options__types">
-                <select id="select-event-category" name="{{ constant('PARAMETER_EVENT_CATEGORY', search) }}" class="form">
-                    <option value=""{% if not search.eventCategory %} selected="selected"{% endif %}>Tous les événements</option>
+                <select id="select-event-category" name="{{ constant('PARAMETER_EVENT_CATEGORY', search) }}"
+                        class="form">
+                    <option value=""{% if not search.eventCategory %} selected="selected"{% endif %}>Tous les
+                        événements
+                    </option>
                     <option value="{{ constant('TYPE_CITIZEN_ACTIONS', search) }}">Actions citoyennes</option>
                     {% for category in event_categories %}
```
[the remark this answers is not in the document]
Comment (Raulnet) [reply templates/search/_filters.html.twig]: :+1: 
edited

File: templates/search/_filters.html.twig (hunk 10 above)
Comment (Raulnet) [reply templates/search/_filters.html.twig]: removed

File: src/Entity/EventGroupCategory.php [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public function addEventCategory(EventCategory $eventCategory): self
+    {
+        if (!$this->eventCategories->contains($eventCategory)) {
+            $this->eventCategories[] = $eventCategory;
+            $eventCategory->setEventGroupCategory($this);
+        }
+
+        return $this;
```
Comment (Remg) [inline src/Entity/EventGroupCategory.php]: is this useful ? 
we don't use fluent interface on this project

File: src/Entity/EventGroupCategory.php [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public function removeEventCategory(EventCategory $eventCategory): self
+    {
+        $this->eventCategories->removeElement($eventCategory);
+        if ($eventCategory->getEventGroupCategory() === $this) {
+            $eventCategory->setEventGroupCategory(null);
+        }
+
+        return $this;
```
Comment (Remg) [inline src/Entity/EventGroupCategory.php]: same thing here, we don't use fluent interface

File: src/Entity/EventGroupCategory.php (hunk 15 above)
Comment (Raulnet) [reply src/Entity/EventGroupCategory.php]: removed

File: src/Entity/EventGroupCategory.php (hunk 16 above)
Comment (Raulnet) [reply src/Entity/EventGroupCategory.php]: removed

File: src/Form/EventGroupCategoryType.php [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+class EventGroupCategoryType extends AbstractType
+{
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults([
+            'class' => EventCategory::class,
+            'choice_label' => 'name',
+            'query_builder' => function (EntityRepository $er) {
+                return $er->createQueryBuilder('ec')
+                    ->leftJoin('ec.eventGroupCategory', 'egc')
```
Comment (ottaviano) [inline src/Form/EventGroupCategoryType.php]: why left join here if the relation is mandatory?

File: src/Form/EventGroupCategoryType.php [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+                    ->where('ec.status = :status')
+                    ->andWhere('egc.status = :status')
+                    ->orderBy('ec.eventGroupCategory', 'ASC')
+                    ->addOrderBy('ec.name', 'ASC')->setParameters([
```
Comment (ottaviano) [inline src/Form/EventGroupCategoryType.php]: CS: `->call1()->call2()`

File: src/Form/EventGroupCategoryType.php [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+                        'status' => EventCategory::ENABLED,
+                    ]);
+            },
+            'group_by' => function ($category) {
+                /** @var EventCategory $category */
+                if ($category->getEventGroupCategory()) {
```
Comment (ottaviano) [inline src/Form/EventGroupCategoryType.php]: why this test if the category group cannot be null ?

File: src/Entity/EventCategory.php [hunk 20]
```diff
@@ -22,4 +22,19 @@
  */
 class EventCategory extends BaseEventCategory
 {
+    /**
+     * @ORM\ManyToOne(targetEntity="AppBundle\Entity\EventGroupCategory", inversedBy="eventCategories")
+     * @ORM\JoinColumn(name="event_group_category_id", referencedColumnName="id", nullable=false)
+     */
```
Comment (ottaviano) [inline src/Entity/EventCategory.php]: 1) what about validation (`Assert`)?
2) why do you set `name` and `refColName` your self?

File: src/Entity/EventCategory.php [hunk 21]
```diff
@@ -22,4 +22,19 @@
...
+    private $eventGroupCategory;
+
+    public function getEventGroupCategory(): ?EventGroupCategory
```
Comment (ottaviano) [inline src/Entity/EventCategory.php]: why `?`?

File: templates/search/_filters.html.twig [hunk 22]
```diff
@@ -5,10 +5,14 @@
         {% if search.isTypeEvents() %}
             <div class="search__bar__options__types">
                 <select id="select-event-category" name="{{ constant('PARAMETER_EVENT_CATEGORY', search) }}" class="form">
-                    <option value=""{% if not search.eventCategory %} selected="selected"{% endif %}>Tous les événements</option>
+                    <option value=""{% if not search.eventCategory %} selected="selected"{% endif %}>Tous les événements </option>
```
Comment (ottaviano) [inline templates/search/_filters.html.twig]: why this change?

File: tests/Controller/EnMarche/CommitteeManagerControllerTest.php [hunk 23]
```diff
@@ -349,7 +349,7 @@ public function testCommitteeHostCanPublishNewEventWithTimeZone()
 
         $this->assertResponseStatusCode(Response::HTTP_OK, $this->client->getResponse());
 
-        $eventCategory = $this->getEventCategoryIdForName(LoadEventCategoryData::LEGACY_EVENT_CATEGORIES['CE003']);
+        $eventCategory = $this->getEventCategoryIdForName(LoadEventCategoryData::LEGACY_EVENT_CATEGORIES['event-category-3']['name']);
```
Comment (ottaviano) [inline tests/Controller/EnMarche/CommitteeManagerControllerTest.php]: need rollback too?

File: tests/Controller/EnMarche/EventManagerControllerTest.php [hunk 24]
```diff
@@ -83,7 +83,7 @@ public function testOrganizerCanEditEvent()
             'committee_event' => [
                 'name' => 'écologie, débatons-en !',
                 'description' => 'Cette journée sera consacrée à un grand débat sur la question écologique.',
-                'category' => $this->getEventCategoryIdForName(LoadEventCategoryData::LEGACY_EVENT_CATEGORIES['CE003']),
+                'category' => $this->getEventCategoryIdForName(LoadEventCategoryData::LEGACY_EVENT_CATEGORIES['event-category-3']['name']),
```
Comment (ottaviano) [inline tests/Controller/EnMarche/EventManagerControllerTest.php]: same?

File: src/Entity/EventCategory.php (hunk 20 above)
Comment (Raulnet) [reply src/Entity/EventCategory.php]: Assert\NotBlank Added, naming removed.

File: src/Entity/EventCategory.php (hunk 21 above)
Comment (Raulnet) [reply src/Entity/EventCategory.php]: form the form the groupCategory is not preset

File: src/Form/EventGroupCategoryType.php (hunk 17 above)
Comment (Raulnet) [reply src/Form/EventGroupCategoryType.php]: edited

File: src/Form/EventGroupCategoryType.php (hunk 18 above)
Comment (Raulnet) [reply src/Form/EventGroupCategoryType.php]: indented

File: src/Form/EventGroupCategoryType.php (hunk 19 above)
Comment (Raulnet) [reply src/Form/EventGroupCategoryType.php]: removed

File: templates/search/_filters.html.twig (hunk 22 above)
Comment (Raulnet) [reply templates/search/_filters.html.twig]: removed

File: tests/Controller/EnMarche/CommitteeManagerControllerTest.php (hunk 23 above)
Comment (Raulnet) [reply tests/Controller/EnMarche/CommitteeManagerControllerTest.php]: removed on commit fix testu

File: tests/Controller/EnMarche/EventManagerControllerTest.php (hunk 24 above)
Comment (Raulnet) [reply tests/Controller/EnMarche/EventManagerControllerTest.php]: edited on commit fix testu

File: src/DataFixtures/ORM/LoadEventCategoryData.php [hunk 25]
```diff
@@ -21,16 +22,42 @@ class LoadEventCategoryData implements FixtureInterface
...
     ];
-    const HIDDEN_CATEGORY_NAME = 'Catégorie masquée';
+
+    public const LEGACY_EVENT_CATEGORIES_GROUPED = [
+        'CE012' => ['name' => 'ancrage local', 'status' => BaseEventCategory::ENABLED, 'group' => 'event-group-category-1'],
+        'CE013' => ['name' => 'projets citoyens', 'status' => BaseEventCategory::ENABLED, 'group' => 'event-group-category-1'],
+        'CE014' => ['name' => 'Un An', 'status' => BaseEventCategory::ENABLED, 'group' => 'event-group-category-1'],
+        'CE015' => ['name' => 'Débat', 'status' => BaseEventCategory::ENABLED, 'group' => 'event-group-category-2'],
+        'CE016' => ['name' => 'Catégorie masquée', 'status' => BaseEventCategory::DISABLED, 'group' => 'event-group-category-0'],
+        ];
```
Comment (Remg) [inline src/DataFixtures/ORM/LoadEventCategoryData.php]: extra indentation

File: src/Form/EventGroupCategoryType.php [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class EventGroupCategoryType extends AbstractType
+{
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults([
+            'class' => EventCategory::class,
+            'choice_label' => 'name',
+            'query_builder' => function (EntityRepository $er) {
```
Comment (Remg) [inline src/Form/EventGroupCategoryType.php]: This logic should take place in a repository imho..

how about you typehint with the specific repository, and then use a custom method of that repository ?
e.g.: 
```
'query_builder' => function (EventCategoryRepository $repository) {
    return $repository->createQueryBuilderForMyEpicFormChoices();
}
```
What do you think?

File: templates/search/_filters.html.twig (hunk 22 above)
[the remark this answers is not in the document]
Comment (Raulnet) [reply templates/search/_filters.html.twig]: GG

File: src/DataFixtures/ORM/LoadEventCategoryData.php (hunk 25 above)
Comment (Raulnet) [reply src/DataFixtures/ORM/LoadEventCategoryData.php]: indented

File: src/Form/EventGroupCategoryType.php (hunk 26 above)
Comment (Raulnet) [reply src/Form/EventGroupCategoryType.php]: edited :+1:
